browser/brave_news/brave_news_tab_helper.cc

```diff
@@ -89,48 +104,103 @@ bool BraveNewsTabHelper::IsSubscribed(const FeedDetails& feed_details) {
+  // the title should be.
+  if (it == rss_page_feeds_.end()) {
+    return "";
+  }
+
+  // Return our cached title, if we have one.
+  if (!it->title.empty()) {
+    return it->title;
+  }
+
+  // If we have an entry, but don't have a title for it, we should request the
+  // feed (if we haven't already). Observers will be notified once the feed
+  // resolves and we extract a title.
+  if (!it->requested_feed) {
+    const auto url = web_contents()->GetLastCommittedURL();
+    controller_->FindFeeds(
+        feed_url,
+        base::BindOnce(&BraveNewsTabHelper::OnFoundFeedData,
+                       weak_ptr_factory_.GetWeakPtr(), feed_url, url));
+    it->requested_feed = true;
   }
-  return false;
+
+  // In the meantime, use the feed_url as the title. Observers will be
+  // notified when we receive a title.
+  return feed_url.spec();
 }
 
-void BraveNewsTabHelper::ToggleSubscription(const FeedDetails& feed_details) {
-  bool subscribed = IsSubscribed(feed_details);
-  auto* publisher = controller_->publisher_controller()->GetPublisherForFeed(
-      feed_details.feed_url);
+void BraveNewsTabHelper::ToggleSubscription(const GURL& feed_url) {
+  bool subscribed = IsSubscribed(feed_url);
+  auto* publisher =
+      controller_->publisher_controller()->GetPublisherForFeed(feed_url);
   if (publisher) {
     controller_->SetPublisherPref(
         publisher->publisher_id, subscribed
                                      ? brave_news::mojom::UserEnabled::DISABLED
                                      : brave_news::mojom::UserEnabled::ENABLED);
   } else if (!subscribed) {
-    controller_->SubscribeToNewDirectFeed(feed_details.feed_url,
-                                          base::DoNothing());
+    controller_->SubscribeToNewDirectFeed(feed_url, base::DoNothing());
   }
 }
 
 void BraveNewsTabHelper::OnReceivedRssUrls(const GURL& site_url,
                                            std::vector<GURL> feed_urls) {
-  if (site_url != GetWebContents().GetLastCommittedURL())
+  if (site_url != GetWebContents().GetLastCommittedURL()) {
     return;
+  }
 
+  // First things first, we just store the urls. This let's us know that we have
+  // feeds, so we should show the button.
   for (const auto& url : feed_urls) {
-    controller_->FindFeeds(
-        url, base::BindOnce(&BraveNewsTabHelper::OnFoundFeeds,
-                            weak_ptr_factory_.GetWeakPtr(), site_url));
+    rss_page_feeds_.push_back({.feed_url = url, .title = ""});
   }
+
+  AvailableFeedsChanged();
 }
 
-void BraveNewsTabHelper::OnFoundFeeds(
+void BraveNewsTabHelper::OnFoundFeedData(
+    const GURL& feed_url,
     const GURL& site_url,
     std::vector<brave_news::mojom::FeedSearchResultItemPtr> feeds) {
-  if (site_url != GetWebContents().GetLastCommittedURL())
+  if (site_url != GetWebContents().GetLastCommittedURL()) {
     return;
+  }
+
+  if (feeds.empty()) {
+    rss_page_feeds_.erase(base::ranges::remove(rss_page_feeds_, feed_url,
```

Comment:
this usage is incorrect and will only at most one item. You need `rss_page_feeds_.end()` as the last param of `erase`

Reply:
also there should be a test covering this